tests/Spec/Types.hs

```diff
@@ -74,25 +94,128 @@ instance Arbitrary M.Challenge where
+  arbitrary = M.AAGUID <$> bytes 16
+
+bytes :: Int -> Gen BS.ByteString
+bytes n = BS.pack <$> replicateM n arbitrary
 
 instance Arbitrary M.CredentialId where
   arbitrary = M.CredentialId <$> arbitrary
 
 instance Arbitrary M.AuthenticatorExtensionOutputs where
   arbitrary = pure M.AuthenticatorExtensionOutputs {}
+
+instance Arbitrary M.RpId where
+  arbitrary = M.RpId <$> arbitrary
+
+instance Arbitrary M.RelyingPartyName where
+  arbitrary = M.RelyingPartyName <$> arbitrary
+
+instance Arbitrary M.PublicKeyCredentialRpEntity where
+  arbitrary =
+    M.PublicKeyCredentialRpEntity
+      <$> arbitrary
+      <*> arbitrary
+
+instance Arbitrary M.UserHandle where
+  arbitrary = M.UserHandle <$> arbitrary
+
+instance Arbitrary M.UserAccountDisplayName where
+  arbitrary = M.UserAccountDisplayName <$> arbitrary
+
+instance Arbitrary M.UserAccountName where
+  arbitrary = M.UserAccountName <$> arbitrary
+
+instance Arbitrary M.PublicKeyCredentialUserEntity where
+  arbitrary =
+    M.PublicKeyCredentialUserEntity
+      <$> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+
+instance Arbitrary M.Timeout where
+  arbitrary = M.Timeout <$> arbitrary
+
+instance Arbitrary M.PublicKeyCredentialDescriptor where
+  arbitrary =
+    M.PublicKeyCredentialDescriptor M.PublicKeyCredentialTypePublicKey
+      <$> arbitrary
+      <*> liftArbitrary shuffledSubset
+
+instance Arbitrary M.AuthenticatorSelectionCriteria where
+  arbitrary =
+    M.AuthenticatorSelectionCriteria
+      <$> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+
+instance Arbitrary M.AuthenticationExtensionsClientInputs where
+  arbitrary = pure M.AuthenticationExtensionsClientInputs
+
+instance Arbitrary (M.PublicKeyCredentialOptions 'M.Create) where
+  arbitrary =
+    M.PublicKeyCredentialCreationOptions
+      <$> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> parameters
+      <*> arbitrary
+      <*> resize 4 arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+
+instance Arbitrary (M.PublicKeyCredentialOptions 'M.Get) where
+  arbitrary =
+    M.PublicKeyCredentialRequestOptions
+      <$> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+      <*> resize 4 arbitrary
+      <*> arbitrary
+      <*> arbitrary
+
+instance Arbitrary M.AuthenticationExtensionsClientOutputs where
+  arbitrary = pure M.AuthenticationExtensionsClientOutputs
+
+instance Arbitrary (M.PublicKeyCredential 'M.Create 'False) where
+  arbitrary =
+    M.PublicKeyCredential
+      <$> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+
+instance Arbitrary (M.PublicKeyCredential 'M.Get 'False) where
+  arbitrary =
+    M.PublicKeyCredential
+      <$> arbitrary
+      <*> arbitrary
+      <*> arbitrary
+
+shuffledSubset :: (Ord a, Bounded a, Enum a) => Gen [a]
+shuffledSubset = subset >>= shuffle . Set.toList
+
+subset :: (Ord a, Bounded a, Enum a) => Gen (Set a)
+subset = Set.fromList <$> sublistOf (Set.toList completeSet)
```

Comment:
I would assume `toList` produces a list with some ordering, meaning that this subset function doesn't produce all valid subsets. Perhaps shuffling the list of the the completeSet before creating a sublist would work? Or perhaps we don't even want all possible subsets, but then the function should be renamed.

Reply:
`sublistOf` generates a [subsequence](https://en.wikipedia.org/wiki/Subsequence) of the given list, which is what we need here.